pytest_httpserver/httpserver.py

```diff
@@ -693,6 +677,219 @@ def create_matcher(self, *args, **kwargs) -> RequestMatcher:
+        the thread is stopped.
+
+        If the sever is already running :py:class:`HTTPServerError` will be raised. If you are
+        unsure, call :py:meth:`is_running` first.
+
+        There's a context interface of this class which stops the server when the context block ends.
+        """
+        if self.is_running():
+            raise HTTPServerError("Server is already running")
+
+        self.server = make_server(self.host, self.port, self.application, ssl_context=self.ssl_context)
+        self.port = self.server.port  # Update port (needed if `port` was set to 0)
+        self.server_thread = threading.Thread(target=self.thread_target)
+        self.server_thread.start()
+
+    def stop(self):
+        """
+        Stop the running server.
+
+        Notifies the server thread about the intention of the stopping, and the thread will
+        terminate itself. This needs about 0.5 seconds in worst case.
+
+        Only a running server can be stopped. If the sever is not running, :py:class`HTTPServerError`
+        will be raised.
+        """
+        if not self.is_running():
+            raise HTTPServerError("Server is not running")
+        self.server.shutdown()
+        self.server_thread.join()
+        self.server = None
+        self.server_thread = None
+
+    def add_assertion(self, obj):
+        """
+        Add a new assertion
+
+        Assertions can be added here, and when :py:meth:`check_assertions` is called,
+        it will raise AssertionError for pytest with the object specified here.
+
+        :param obj: An AssertionError, or an object which will be passed to an AssertionError.
+        """
+        self.assertions.append(obj)
+
+    def check(self):
+        """
+        Raises AssertionError or Errors raised in handlers.
+
+        Runs both :py:meth:`check_assertions` and :py:meth:`check_handler_errors`
+        """
+        self.check_assertions()
+        self.check_handler_errors()
+
+    def check_assertions(self):
+        """
+        Raise AssertionError when at least one assertion added
+
+        The first assertion added by :py:meth:`add_assertion` will be raised and
+        it will be removed from the list.
+
+        This method can be useful to get some insights into the errors happened in
+        the sever, and to have a proper error reporting in pytest.
+        """
+
+        if self.assertions:
+            assertion = self.assertions.pop(0)
+            if isinstance(assertion, AssertionError):
+                raise assertion
+
+            raise AssertionError(assertion)
+
+    def check_handler_errors(self):
+        """
+        Re-Raises any errors caused in request handlers
+
+        The first error raised by a handler will be re-raised here, and then
+        removed from the list.
+        """
+        if self.handler_errors:
+            raise self.handler_errors.pop(0)
+
+    def respond_nohandler(self, request: Request, extra_message: str = ""):
```

Comment:
out of curiosity, is extra_message specified somewhere where `respond_nohandler` it is called?

Reply:
In the derived class, it is only for eliminating the coding style error that the signature is different from the base class one.
The base class one is called with the formatted matchers.